core/src/main/java/tc/oc/pgm/filters/dynamic/FilterMatchModule.java

```diff
@@ -0,0 +1,605 @@
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.filter.Filter;
+import tc.oc.pgm.api.filter.query.Query;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.match.Tickable;
+import tc.oc.pgm.api.match.event.MatchLoadEvent;
+import tc.oc.pgm.api.party.Party;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.player.ParticipantState;
+import tc.oc.pgm.api.player.event.MatchPlayerDeathEvent;
+import tc.oc.pgm.api.time.Tick;
+import tc.oc.pgm.events.ListenerScope;
+import tc.oc.pgm.events.PlayerPartyChangeEvent;
+import tc.oc.pgm.filters.TimeFilter;
+import tc.oc.pgm.flag.event.FlagStateChangeEvent;
+import tc.oc.pgm.util.MapUtils;
+import tc.oc.pgm.util.event.PlayerCoarseMoveEvent;
+
+/**
+ * Handler of dynamic filter magic &#x1F52E;.
+ *
+ * <p>A {@link Filter} can have the possibility of being dynamic or not, but the match module
+ * decides whether to use a filter dynamically or not. In this sense the docs can be a little
+ * misleading when thinking about dynamic filters programmatically. The docs say "the [dynamic]
+ * filter will notify the module when it's time to do something". In reality this {@link
+ * FilterMatchModule} will "notify" the modules using registered {@link FilterListener}s to execute
+ * code in the module.
+ *
+ * <p>When registering {@link FilterListener}s we use the words "rise" and "fall" to describe the
+ * states of a dynamic filter we want to listen for. The relevant methods are explained in {@link
+ * FilterDispatcher}.
+ *
+ * @see #onChange(Class, Filter, FilterListener)
+ * @see #onRise(Class, Filter, Consumer)
+ * @see #onFall(Class, Filter, Consumer)
+ */
+@ListenerScope(MatchScope.LOADED)
+public class FilterMatchModule implements MatchModule, FilterDispatcher, Tickable, Listener {
+
+  private final Match match;
+  private final List<Class<? extends Event>> listeningFor = new LinkedList<>();
+  private final PriorityQueue<TimeFilter> timeFilterQueue = new PriorityQueue<>();
+
+  // MethodHandle shinanigans
+  private final MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+  private final String[] filterableGetterNames = new String[] {"getPlayer", "getParty", "getMatch"};
+  private final MethodType[] filterableMethodTypes =
+      new MethodType[] {
+        MethodType.methodType(MatchPlayer.class),
+        MethodType.methodType(Party.class),
+        MethodType.methodType(Match.class)
+      };
+
+  // #getActor is mostly for SportPaper (PlayerAction).
+  // #getPlayer covers events which does not provide Entity
+  // as the lower boundary
+  private final String[] entityGetterNames = new String[] {"getPlayer", "getActor", "getActor"};
+  private final MethodType playerMethodType = MethodType.methodType(Player.class);
+  private final MethodType[] entityMethodTypes =
+      new MethodType[] {
+        this.playerMethodType, this.playerMethodType, MethodType.methodType(Entity.class)
+      };
```

Comment:
This is a 600 line class where half of it is about doing reflection on events to get player or actor

Outsource all of that logic to somewhere else and keep this to just the filter stuff, whenever you need to get this stuff call the util.

Make that util static and reuse all of the event -> method stuff, you don't need to create that for each match

Reply:
Added bonus of moving out this stuff is that now cached handles persist over different matches.